src/storage/src/storage/bidi/worker.rs

```diff
@@ -0,0 +1,733 @@
+        assert!(result.is_ok(), "{result:?}");
+        Ok(())
+    }
+
+    #[test_case(true)]
+    #[test_case(false)]
+    #[tokio::test]
+    async fn run_bad_response(range_end: bool) -> anyhow::Result<()> {
+        let (request_tx, _request_rx) = tokio::sync::mpsc::channel(1);
+        let (response_tx, response_rx) = mock_stream();
+        let (_tx, rx) = tokio::sync::mpsc::channel(1);
+        let connection = Connection::new(request_tx, response_rx);
+
+        // Simulate a response for an unexpected read id.
+        let response = BidiReadObjectResponse {
+            object_data_ranges: vec![ObjectRangeData {
+                read_range: Some(proto_range_id(0, 100, -123)),
+                range_end,
+                ..ObjectRangeData::default()
+            }],
+            ..BidiReadObjectResponse::default()
+        };
+        response_tx.send(Ok(response)).await?;
+        let mut mock = MockTestClient::new();
+        mock.expect_start().never();
+
+        let connector = mock_connector(mock);
+        let worker = Worker::new(connector);
+        let err = worker.run(connection, rx).await.unwrap_err();
+        assert!(err.is_transport(), "{err:?}");
+        let source = err.source().and_then(|e| e.downcast_ref::<ReadError>());
+        assert!(
+            matches!(source, Some(ReadError::UnknownBidiRangeId(r)) if *r == -123),
+            "{err:?}"
+        );
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn run_reconnect() -> anyhow::Result<()> {
+        let (request_tx, _request_rx) = tokio::sync::mpsc::channel(1);
+        let (response_tx, response_rx) = mock_stream();
+        let (_tx, rx) = tokio::sync::mpsc::channel(1);
+        let connection = Connection::new(request_tx, response_rx);
+
+        // Simulate a redirect response.
+        response_tx
+            .send(Err(redirect_status("redirect-01")))
+            .await?;
+        let mut mock = MockTestClient::new();
+        mock.expect_start()
+            .return_once(|_, _, _, _, _, _| Err(permanent_error()));
+
+        let connector = mock_connector(mock);
+        let worker = Worker::new(connector);
+        let err = worker.run(connection, rx).await.unwrap_err();
+        assert_eq!(err.status(), permanent_error().status());
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn run_stop_on_closed_requests() -> anyhow::Result<()> {
+        let (request_tx, _request_rx) = tokio::sync::mpsc::channel(1);
+        let (_response_tx, response_rx) = mock_stream();
+        let (tx, rx) = tokio::sync::mpsc::channel(1);
+        let connection = Connection::new(request_tx, response_rx);
+
+        let mut mock = MockTestClient::new();
+        mock.expect_start().never();
+
+        let connector = mock_connector(mock);
+        let worker = Worker::new(connector);
+        drop(tx);
+        worker.run(connection, rx).await?;
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn run_partial_read() -> anyhow::Result<()> {
```

Comment:
This test and run_full_read are very similar. It looks like the segment is different and the full read checks that the we expect to get None as a response after.

Can we combine these to share the code? Or possible add the comment about not expecting the stream to be done (would be great if we could confirm not None but that seems hard).